An implemenation of the ITU-R BS.1770-4 loudness recommendation

bshall opened the pull request:
I took a stab at implementing the ITU-R BS.1770-4 loudness recommendation (closes #1205). To give some more details:
- I've implemented K-weighting following csteinmetz1 instead of BrechtDeMan since it fit well with torchaudio's already implemented filters (`treble_biquad` and `highpass_biquad`).
- I've added four audio files to test compliance with the recommendation. These are linked in [this pdf](https://www.itu.int/dms_pub/itu-r/opb/rep/R-REP-BS.2217-2-2016-PDF-E.pdf). There are many more test files there but I didn't want to bog down the assets directory with too many files. Let me know if I should add or remove anything.
- I've kept many of the constant internal to the function (e.g. the block duration, overlap, and the absolute threshold gamma). I'm not sure if these should be exposed in the signature.
- I've implemented support for up to 5 channels (following both csteinmetz1 and BrechtDeMan). The recommendation includes weights for up to 24 channels. Is there any convention for how many channels to support?

I hope this is helpful! looking forward to hearing from you.

mthrok reviewed (commented):
Hi @bshall 

Thanks for the PR. 
Quick questions before looking into the PR:
1. Where are the reference files from? I can get the article https://www.itu.int/pub/R-REP-BS.2217-2-2016, but cannot find links to ~the data~ (EDIT: PDF contains the link to the data) nor the license information. If the license is unclear, we cannot checkin the files. In that case, we need to work around by fetching data from a web at test runtime.
2. Currently torchaudio main branch is 14.83 MB. The assets will double the size of main branch. Is there a way to reduce the asset size?

bshall commented:
Hi @mthrok, I did some digging around and couldn't find a license on the ITU website so I rewrote the tests to download the files at runtime. I created a new test class but can add the tests to the `Functional` class if that's preferable. I've also squashed the commits so there isn't any history of the files in the branch. Also, this way we don't have to worry about increasing the size of assets :)

bshall commented:
Hi @mthrok, just wanted to follow up on this pull request since there hasn't been an update in a little while. I've rebased on main so it should be up to date.

mthrok commented:
Hi @bshall 

Sorry I had been busy regarding the recent release. I will have time to review this soon.
One thing about the testing, one of the restrictions on unit test is that the test should not rely on external resource. Since now that the assets are fetched on-the-fly, can we move the test to integration directory?

mthrok commented:
@carolineechen Can you help review?

carolineechen reviewed (commented):
Thanks for working on this! Left a couple of comments, mostly small code suggestions and the two main things are
- moving the test to integration_tests bc of reliance on external source
- adding this function to transforms as well

bshall commented on torchaudio/functional/functional.py:
```diff
@@ -1602,6 +1604,66 @@ def edit_distance(seq1: Sequence, seq2: Sequence) -> int:
+        sample_rate (int): sampling rate of the waveform, e.g. 44100 (Hz)
+
+    Returns:
+        Tensor: loudness estimates (LKFS)
```
Currently the function works on batched inputs as well so I'm returning a Tensor of LKFS values to deal with this case.

bshall commented on torchaudio/functional/functional.py:
```diff
@@ -1602,6 +1604,66 @@ def edit_distance(seq1: Sequence, seq2: Sequence) -> int:
+    if waveform.size(-2) > 5:
+        raise ValueError("Only up to 5 channels are supported.")
+
+    gate_duration: float = 0.4
+    overlap: float = 0.75
+    gamma_abs: float = -70.0
```
I'm not sure if there's a potential use case for adjusting any of the parameters. I've always just stuck to the settings in the recommendation and I'm sure that covers most of the usage. On the other hand it's pretty easy to expose these values so maybe it's worth doing just for the flexibility?

carolineechen replied:
got it, in that case let's keep it defined in the function for now. and if we get a feature request with a use case in the future asking to expose any of the parameters, we can easily address it then.

bshall commented on test/torchaudio_unittest/functional/loudness_compliance_test_impl.py:
```diff
@@ -0,0 +1,63 @@
+class Loudness(TempDirMixin, TestBaseMixin):
+    def download_and_extract_file(self, filename):
+        zippath = self.get_temp_path(filename + ".zip")
+        torch.hub.download_url_to_file(_COMPLIANCE_FILE_URLS[filename], zippath, progress=False)
```
I'm not sure how I'd use download_asset from torchaudio.utils. It looks like the key has to be relative to https://download.pytorch.org/torchaudio/. Let me know if I'm missing something.

carolineechen replied:
this is my bad, we can keep the original `torch.hub.download_url_to_file`, thanks for catching this

carolineechen commented on torchaudio/transforms/_transforms.py:
```diff
@@ -1248,6 +1248,32 @@ def __init__(self, time_mask_param: int, iid_masks: bool = False, p: float = 1.0
+        - https://www.itu.int/rec/R-REC-BS.1770-4-201510-I/en
+    """
+
+    def __init__(self):
```
let's move the `sample_rate` parameter into `__init__`, set `self.sample_rate`, and move the docstrings up too

carolineechen commented on test/integration_tests/loudness_compliance_test_impl.py:
```diff
@@ -0,0 +1,60 @@
+            file.extractall(os.path.dirname(zippath))
+        return self.get_temp_path(filename + ".wav")
+
+    def test_measure_loudness_relative_gate(self):
```
looks like these tests just differ in the file path name and the expected tensor. can you reduce it to one test function and test for the different samples using `parameterized`?

```
parameterized.expand(
    [
        ("1770-2_Comp_RelGateTest", -10.0),
        ("1770-2_Comp_AbsGateTest", -69.5),
         ...
    ]
)
def test_measure_loudness(self, filepath, expected):
    ....
```

carolineechen reviewed (commented):
thanks for the fixes and responses, a few more things

carolineechen commented:
Could you also rebase onto main (`build_docs` job on CircleCI should be resolved), and also add the appropriate docs entry into `functional.rst` and `transforms.rst`?

bshall commented:
Hi @carolineechen, thanks for the feedback! I've made all the requested changes and rebased on main so hopefully everything is in good shape now. Let me know if there's anything else.

carolineechen commented:
Changes look good, thanks for working on this! Integration test on CircleCI is failing though, think you may need to add `expecttest` and `parameterized` to the integration test workflow [here](https://github.com/pytorch/audio/blob/main/.github/workflows/integration-test.yml#L32)

bshall commented:
@carolineechen, I've added `scipy` to the integration test workflow as well. Hopefully that's the last requirement.

carolineechen commented:
@bshall so sorry for all the complications with the Github build/test, there was an upstream change from PyTorch so unittests on older PRs are now breaking, any chance you could rebase again, or I'm happy to take over resolving the CircleCI testing if that is easier for you. thanks again for the contributions!

mthrok commented:
> Changes look good, thanks for working on this! Integration test on CircleCI is failing though, think you may need to add `expecttest` and `parameterized` to the integration test workflow [here](https://github.com/pytorch/audio/blob/main/.github/workflows/integration-test.yml#L32)

@carolineechen We should not be using unit test fixtures in integration test. Integration tests should run without such harness. This PR is okay, but we need to get rid of unit test fixtures from integration test.

bshall commented:
Hi @carolineechen, no problem! I'm happy to see this one through, but if its taking too long for me to make the changes and you want to jump in that's cool. I've rebased so hopefully the tests will pass now and there won't be any other issues to resolve :)

mthrok commented:
@bshall `torchaudio_unittest` module cannot be used in integration test. Can you either remove it or, mark this PR editable by maintainers so that we can push on top of your commits?

bshall commented:
@mthrok and @carolineechen, I've removed the `torchaudio_unittest` module from the tests and also checked "allow edits by maintainers" for the branch. I guess since this is just an integration test we don't need to worry about testing different `dtypes` or `devices`? Not sure how to do that without importing `torchaudio_unittest`.

carolineechen commented:
[stylecheck](https://app.circleci.com/pipelines/github/pytorch/audio/12014/workflows/f31c6f2f-a0dc-45f4-8414-7a624f9d7803/jobs/829527) is failing, could you make that small fix, and then I think the rest looks good. Thanks for the changes!
